link-grammar/linkage/linkage.c

```diff
@@ -875,6 +875,38 @@ const char * linkage_get_word(const Linkage linkage, WordIdx w)
 	return linkage->word[w];
 }
 
+// Need to free the returned value of this function later:
+// the memory pointed to by bare_word
+char * linkage_get_bare_word(const Linkage linkage, WordIdx w)
+{
+	if (!linkage) return NULL;
+	if (linkage->num_words <= w) return NULL; /* bounds-check */
+
+	// first look for unknown word tag
+	char delimiter = '[';
+	char *separator_ptr = strrchr(linkage->word[w], delimiter);
+
+	if(separator_ptr == NULL)
+	{
+	    // look for beginning of word tag
+		delimiter = '.';
+	    separator_ptr = strrchr(linkage->word[w], delimiter);
+	    if(separator_ptr == NULL)
+		{
+			// if no tags are aded, point to the end of the word
+			delimiter = '\0';
+	  		separator_ptr = strrchr(linkage->word[w], delimiter); // if there is no separator in the word
+		}
+	}
+	// copy and return word without the suffix
+	int position = separator_ptr - linkage->word[w];
+	char* bare_word = (char*) malloc((position + 1) * sizeof(char));
+	memcpy(bare_word, linkage->word[w], position);
```

Comment:
alloca allocates on stack. When the stack frame disappears, the pointer is no longer valid.  

malloc is dangerous, because users have a habit of forgetting to free. the LG API tries hard to avoid returning malloced memory.

LG subwords are used to hold morphemes and spell-guessed word-splits. They probably don't do what you think they do, unless you carefully studied the Russian link-grammar, where they are used heavily.  The simplest English example would be "seasond" which might be one word: "seasoned" or it might be two: "seas and"  "sea sand".

The functions that Amir added provide both byte and character offsets. In English, these are the same thing, for non-English, they are not.